add 25720 notes (#539)
bitcoin-core-review-club/website: _posts/2022-08-10-#25720.md

commit:
---

## Notes

- On startup, if a node's headers chain is more than a day behind the current
  time, it will pick one peer to sync headers with until its best headers chain
  is caught up. (See
  [here](https://github.com/bitcoin/bitcoin/blob/92f6461cfd39fff2fc885dd623fa47e7d8d53827/src/net_processing.cpp#L4889-L4913))
  If a node's headers chain is not caught up within `15 min + ~num headers
  missing * 1ms` (either because the chosen peer was not able to or because

>   they purposefully stalled the download), it will disconnect the chosen peer

Reviewer: ```suggestion
  they purposefully stalled the download), it will [disconnect](https://github.com/bitcoin/bitcoin/blob/92f6461cfd39fff2fc885dd623fa47e7d8d53827/src/net_processing.cpp#L5223-L5254) the chosen peer
```

>   for stalling and pick a new peer to sync headers with.

Reviewer: Perhaps useful to mention headers-first approach and that we don't need the blocks to be synced within 15mins, just the headers?

Author: Added a sentence for block download

> - When a node receives a block announcement via an INV during headers sync, it

Reviewer: ```suggestion
- When a node [receives](https://github.com/bitcoin/bitcoin/blob/92f6461cfd39fff2fc885dd623fa47e7d8d53827/src/net_processing.cpp#L3236-L3248) a block announcement via an INV during headers sync, it
```

>   will add all announcing peers as additional headers sync peers. This week's

Reviewer: ```suggestion
  will [add](https://github.com/bitcoin/bitcoin/blob/92f6461cfd39fff2fc885dd623fa47e7d8d53827/src/net_processing.cpp#L3268-L3274) all announcing peers as additional headers sync peers. This week's
```

>   will add all announcing peers as additional headers sync peers. This week's
>   PR changes this logic such that only one of the announcing peers is added for

Reviewer 2: ```suggestion
  will add all announcing peers as additional headers sync peers.
  [PR #25720](https://github.com/bitcoin/bitcoin/pull/25720)
  changes this logic such that only one of the announcing peers is added for
```

  headers sync.

- [BIP 130](https://github.com/bitcoin/bips/blob/master/bip-0130.mediawiki)

>   (implemented in Bitcoin Core since late 2015) introduced the `sendheaders`

Reviewer 2: Maybe link https://github.com/bitcoin/bitcoin/pull/7129 instead, to be more specific?

Author: Now it's linking to both

  message, which indicates that a node prefers receiving block announcements
  via a `headers` message rather than an `inv`.

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
   NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?

> 2. Why do nodes (mostly) receive `inv` block announcements while they are doing

Reviewer: nit: sometimes referred to as "`inv`", sometimes as "INV" - would pick one throughout

Reviewer 2: I prefer lowercase inv as used in pchCommands.
Seems these are all `inv` in the notes now, so resolving.

   initial headers sync, even though they indicated preference for `headers`
   announcements (BIP 130)?

3. Why is bandwidth wasted (during initial headers sync) by adding all peers
   that announce a block to us via an INV as headers sync peers?

> 4. What would your estimate (lower/upper bound) be of how much bandwidth is

Reviewer: ```suggestion
4. What would be your estimate (lower/upper bound) of how much bandwidth is
```

   wasted? (There is no one true answer here, the estimate depends on a couple
   of variables)

5. If we start syncing headers with peers that announce a block to us via an

>    INV, why do we not increase `nSyncStarted` and set `fSyncStarted = true` and

Reviewer: With "why", do you mean the rationale or why it doesn't happen?

Author: I mean why it doesn't happen. It's a question that i asked myself while reviewing, because i think these variables could totally be set at that time.

   `m_headers_sync_timeout`? (Please start by explaining what these variables

>    are used for)

Reviewer: I think it's an easier format (especially live in the review club) to break that up so you can ask the first question first, e.g.:

What's the purpose of `CNodeState`'s members `fSyncStarted` and `m_headers_sync_timeout`, and `PeerManagerImpl::nSyncStarted`? If we start syncing headers with peers that announce a block to us via an
   INV, why do we not increase `nSyncStarted` and set `fSyncStarted = true` and update
   `m_headers_sync_timeout`?
